File: Algorithms/p219_Contains_Duplicate_II/p219_Contains_Duplicate_II.py
class Solution(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        pos = min(len(nums),k+1)
        s = set(nums[:pos])
        if len(s)<pos:
            return True
        for i in range(pos,len(nums)):
            # remove before add for cases where nums[i-k-1]==nums[i]
            s.remove(nums[i-k-1])
            s.add(nums[i])
            if len(s)<k+1:
                return True
        return False

File: Algorithms/p219_Contains_Duplicate_II/p219_Contains_Duplicate_II.cpp
class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        int pos = min(int(nums.size()),k+1);
        set<int> s(nums.begin(),nums.begin()+pos);
        if(s.size()<pos) return 1;
        for(int i=pos;i<nums.size();++i){
            // erase before insert for cases where nums[i-k-1]==nums[i]
            s.erase(nums[i-k-1]);
            s.insert(nums[i]);
            if(s.size()<k+1) return 1;
        }
        return 0;
    }
};
